aten/src/ATen/native/TensorShape.cpp

```diff
@@ -3757,22 +3757,59 @@ std::vector<Tensor> unflatten_dense_tensors(const Tensor& flat, TensorList tenso
   return outputs;
 }
 
+
+// Clones a tensor by cloning the underlying storage that it came from,
+// which allows us to replicate the exact strides/storage_offset in the cloned tensor.
+// Note [*_scatter ops preserve strides]
+// In order for functionalization to preserve stride correctness, the *_scatter
+// operators that it calls must preserve the striding behavior of their inputs.
+// Specifically, the output of *_scatter(base, mutated_view, ...)
+// should have identical size/stride/storage_offset to "base".
+at::Tensor clone_preserve_strides(const at::Tensor& self) {
+  TORCH_INTERNAL_ASSERT(self.has_storage());
+  // In cases where the input tensor has internal memory overlap, we cannot actually
+  // preserve the strides/storage_offset of the input tensor, because
+  // *_scatter ops will try to copy_() into the cloned tensor.
+  // However, this should **never** show up in functionalized user code;
+  // most aten ops that try to mutate a tensor with internal memory overlap would error anyway.
+  //
+  // The one place that this does come up is in autograd - if there's a select_scatter
+  // in the forward, then autograd will generate one for the backward.
+  // If the input to the select_scatter is grad_output, then this could be an expanded tensor
+  // with internal overlap.
+  if (at::has_internal_overlap(self) == at::MemOverlap::Yes) {
+    return self.clone();
```

Comment:
You sure you don't want to just error here?

Reply:
So... maybe I should investigate this more. There was a test failure for the `select_scatter` opinfo `test_aotdispatch` test. The test itself didn't involve any mutations (so functionalization wasn't manually inserting a `select_scatter` anywhere). But the OpInfo calls `select_scatter` in the forward, which caused `select_scatter` to show up in the backward, where in that test case, the input to it (`grad_output`) had overlapping memory, probably because it came from an `expand()` call.

Reply:
My thought was that this matters less, because... as long as functionalization runs underneath autograd, then autograd won't see any `*_scatter` ops directly, so we won't have to worry about those ops showing up in the backward. It's only really a problem if user code is using `*_scatter` ops, which they (probably) shouldn't be.

Reply:
I guess we could error here. But that would basically make `*_scatter` op calls in user code completely unusable.

I could see that being what we want, anyway? Kind of annoying though, because it's nice to have some op info coverage for the `*_scatter` ops, even if they wouldn't work at all with autograd.

Reply:
OK, the expand thing seems valid. You can keep it as is.